openmls/src/ciphersuite/mod.rs

```diff
@@ -538,6 +549,203 @@ impl Signature {
+        }
+
+        let mut r = decode_scalar(&mut signature_bytes)?;
+        let mut s = decode_scalar(&mut signature_bytes)?;
+
+        // After reading the whole signature, nothing should be left.
+        if !signature_bytes.is_empty() {
+            return Err(SignatureError::DecodingError);
+        }
+
+        let mut out = Vec::with_capacity(2 * P256_SCALAR_LENGTH);
+        out.append(&mut r);
+        out.append(&mut s);
+        Ok(out)
+    }
+
+    /// This function DER encodes a given ECDSA signature consisting of bytes
+    /// representing the concatenated scalars. If the encoding fails, it will
+    /// throw a `SignatureError`.
+    pub(crate) fn der_encode(raw_signature: &[u8]) -> Result<Self, SignatureError> {
+        // A small function to DER encode single scalar.
+        fn encode_scalar(mut scalar: &[u8]) -> Result<Vec<u8>, SignatureError> {
+            // We first prepare the scalar DER encoded as a uint.
+            let mut scalar_uint: Vec<u8> = Vec::with_capacity(P256_SCALAR_LENGTH + 1);
+
+            // Remove prepending zeros
+            let mut msb = scalar
+                .read_u8()
+                .map_err(|_| SignatureError::EncodingError)?;
+            while msb == 0x00 {
+                msb = scalar
+                    .read_u8()
+                    .map_err(|_| SignatureError::EncodingError)?;
+            }
+
+            // If the most significant bit is 1, we have to prepend 0x00 to indicate
+            // that the integer is unsigned.
+            if msb > 0x7F {
+                scalar_uint
+                    .write_u8(0x00)
+                    .map_err(|_| SignatureError::EncodingError)?;
+            };
+
+            scalar_uint
+                .write_u8(msb)
+                .map_err(|_| SignatureError::EncodingError)?;
+            scalar_uint
+                .write_all(scalar)
+                .map_err(|_| SignatureError::EncodingError)?;
+
+            let mut encoded_scalar: Vec<u8> = Vec::with_capacity(P256_SCALAR_LENGTH + 3);
+            // The encoded scalar needs to start with integer tag.
+            encoded_scalar
+                .write_u8(INTEGER_TAG)
+                .map_err(|_| SignatureError::EncodingError)?;
+            // This is safe, as we know that r is at most 33 bytes long.
+            encoded_scalar
+                .write_u8(scalar_uint.len() as u8)
+                .map_err(|_| SignatureError::EncodingError)?;
+            encoded_scalar
+                .write_all(&scalar_uint)
+                .map_err(|_| SignatureError::EncodingError)?;
+
+            Ok(encoded_scalar)
+        }
+
+        // Check overall length
+        if raw_signature.len() != 2 * P256_SCALAR_LENGTH {
+            return Err(SignatureError::EncodingError);
+        }
+
+        // We DER encode the ECDSA signature as per spec, assuming that
+        // `sign` returns two concatenated values (r||s).
+        let r = raw_signature
+            .get(..P256_SCALAR_LENGTH)
```

Comment:
You check the length of `raw_signature` before, so you can just do `raw_signature[..P256_SCALAR_LENGTH]` (or don't do the length check).

Reply:
Well, it's a function that can panic. In that sense, yes. But that's what you get when using array access. I'm fine with the `get` here. But you check the length twice because you already know `raw_signature.len() == 2 * P256_SCALAR_LENGTH`. If you want to avoid the potential for panics here I suggest just to drop the `raw_signature.len() != 2 * P256_SCALAR_LENGTH` check.